File: backend_golang/application_service/usecase/create_todo_usecase.go
package usecase

import (
	"backend_golang/domain/model"
	"backend_golang/domain/repository"
	"context"
)

type CreateTodoUsecaseInput struct {
	Title       string
	Description *string
}

type CreateTodoUsecase interface {
	Handle(ctx context.Context, input CreateTodoUsecaseInput) (model.Todo, error)
}

type createTodoUsecaseImpl struct {
	tr repository.TodoRepository
}

func NewCreateTodoUsecase(tr repository.TodoRepository) CreateTodoUsecase {
	return &createTodoUsecaseImpl{
		tr: tr,
	}
}

func (u *createTodoUsecaseImpl) Handle(ctx context.Context, input CreateTodoUsecaseInput) (model.Todo, error) {
	todo := model.NewTodo(input.Title, input.Description)
	if err := u.tr.Save(ctx, todo); err != nil {
		return model.Todo{}, err
	}
	return todo, nil
}

File: backend_rust/src/application/usecase/create_todo_usecase.rs
use std::sync::Arc;

use serde::Deserialize;
use validator::Validate;

use crate::application::error::UsecaseError;
use crate::domain::model::todo::{Todo, TodoDto};
use crate::domain::repository::todo_repository::TodoRepository;

#[derive(Debug, Deserialize, Validate)]
pub struct CreateTodoInput {
    #[validate(length(min = 1))]
    pub title: String,
    pub description: Option<String>,
}

pub struct CreateTodoUsecase {
    todo_repository: Arc<dyn TodoRepository>,
}

impl CreateTodoUsecase {
    pub fn new(todo_repository: Arc<dyn TodoRepository>) -> Self {
        Self { todo_repository }
    }

    pub async fn execute(&self, input: CreateTodoInput) -> Result<TodoDto, UsecaseError> {
        let todo = Todo::new(input.title, input.description);
        self.todo_repository.save(&todo).await?;
        Ok(todo.to_dto())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::application::usecase::test_support::InMemoryTodoRepository;
    use uuid::Uuid;

    #[tokio::test]
    async fn create_persists_todo() {
        let repository = Arc::new(InMemoryTodoRepository::default());
        let usecase = CreateTodoUsecase::new(repository.clone());

        let result = usecase
            .execute(CreateTodoInput {
                title: "in-memory".to_string(),
                description: Some("desc".to_string()),
            })
            .await
            .expect("create todo");

        let stored = repository
            .find_by_id(&Uuid::parse_str(&result.id).unwrap())
            .await
            .expect("todo saved");
        assert_eq!(stored.title(), "in-memory");
        assert_eq!(stored.description(), Some(&"desc".to_string()));
        assert!(!stored.completed());
    }
}
